src/lib/form-field/form-field.ts

```diff
@@ -0,0 +1,382 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {
+  AfterContentChecked,
+  AfterContentInit,
+  AfterViewInit,
+  ChangeDetectionStrategy,
+  ChangeDetectorRef,
+  Component,
+  ContentChild,
+  ContentChildren,
+  Directive,
+  ElementRef,
+  Inject,
+  Input,
+  Optional,
+  QueryList,
+  ViewChild,
+  ViewEncapsulation,
+} from '@angular/core';
+import {animate, state, style, transition, trigger} from '@angular/animations';
+import {coerceBooleanProperty} from '../core';
+import {NgControl} from '@angular/forms';
+import {
+  getMdFormFieldDuplicatedHintError,
+  getMdFormFieldMissingControlError,
+  getMdFormFieldPlaceholderConflictError,
+} from './form-field-errors';
+import {
+  FloatPlaceholderType,
+  MD_PLACEHOLDER_GLOBAL_OPTIONS,
+  PlaceholderOptions
+} from '../core/placeholder/placeholder-options';
+import {startWith} from '@angular/cdk/rxjs';
+import {Observable} from 'rxjs/Observable';
+
+let nextUniqueId = 0;
+
+
+/**
+ * The placeholder directive. The content can declare this to implement more
+ * complex placeholders.
+ */
+@Directive({
+  selector: 'md-placeholder, mat-placeholder'
+})
+export class MdPlaceholder {}
+
+
+/** Hint text to be shown underneath the form field control. */
+@Directive({
+  selector: 'md-hint, mat-hint',
+  host: {
+    'class': 'mat-hint',
+    '[class.mat-right]': 'align == "end"',
+    '[attr.id]': 'id',
+  }
+})
+export class MdHint {
+  /** Whether to align the hint label at the start or end of the line. */
+  @Input() align: 'start' | 'end' = 'start';
+
+  /** Unique ID for the hint. Used for the aria-describedby on the form field control. */
+  @Input() id: string = `md-hint-${nextUniqueId++}`;
+}
+
+
+/** Single error message to be shown underneath the form field. */
+@Directive({
+  selector: 'md-error, mat-error',
+  host: {
+    'class': 'mat-error',
+    'role': 'alert',
+    '[attr.id]': 'id',
+  }
+})
+export class MdError {
+  @Input() id: string = `md-error-${nextUniqueId++}`;
+}
+
+
+/** Prefix to be placed the the front of the form field. */
+@Directive({
+  selector: '[mdPrefix], [matPrefix]',
+})
+export class MdPrefix {}
+
+
+/** Suffix to be placed at the end of the form field. */
+@Directive({
+  selector: '[mdSuffix], [matSuffix]',
+})
+export class MdSuffix {}
+
+
+/** An interface which allows a control to work inside of a `MdFormField`. */
+export abstract class MdFormFieldControl {
```

Comment:
I did this because I need to use the class itself as the injection token so it can be used with `@ViewChild` (something like `InjectionToken<MdFormFieldControl>` causes a type error). It doesn't have to be abstract though, I could use a concrete class with no-op methods.